### ISSUE TITLE
Flutter "in_app_purchase" plugin does not provide any token after completing or restoring the purchases. On the other hand found a link where it was said there will be a token required for sent to back end after in app purchase complete in the application, this is the link : "https://gist.github.com/jamesmosier/8a0d3aa31129f00f92963180376b979f"

### ISSUE THREAD

**Issue by joy-eiosys** [flutter/flutter#95309]
After completing the purchase getting this data in for of JSON : 
(No token is available for sending in the backend)
{
    "purchaseID":"Combination of string and int",
    "productID":"1",
    "transactionDate":"2021-12-15",
    "verificationData":{
        "localVerificationData":"String",
        "serverVerificationData":"String",
        "source":"String"
    }
}

**Comment by darshankawar** [comment]
@joy-eiosys 
Can you provide `flutter doctor -v`, plugin version and a [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) along with steps to replicate this behavior ?

**Comment by joy-eiosys** [comment]
> @joy-eiosys Can you provide `flutter doctor -v`, plugin version and a [complete minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) along with steps to replicate this behavior ?

in_app_purchase: ^1.0.6

Flutter doctor -v :

[√] Flutter (Channel stable, 2.5.3, on Microsoft Windows [Version 10.0.19042.1348], locale en-IN)
    • Flutter version 2.5.3 at C:\flutter_src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (9 weeks ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\JOY\AppData\Local\Android\sdk
    X cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    X Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio      
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.63.0)
    • VS Code at C:\Users\JOY\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.29.0

[√] Connected device (3 available)
    • SM G570F (mobile) • 192.168.0.105:5000 • android-arm    • Android 8.0.0 (API 26)
    • Chrome (web)      • chrome             • web-javascript • Google Chrome 96.0.4664.93
    • Edge (web)        • edge               • web-javascript • Microsoft Edge 96.0.1054.53

! Doctor found issues in 1 category.

minimal reproducible code sample:

StreamSubscription<List<PurchaseDetails>> _subscriptions;

/// In app purchase plugin reference
  InAppPurchase _inAppPurchase = InAppPurchase.instance;

  @override
  void initState() {
    /// Stream of updated purchase
    final Stream<List<PurchaseDetails>> purchaseUpdated =
        _inAppPurchase.purchaseStream;

    /// _subscriptions = purchaseUpdated.listen((purchaseDetailsList) {
      
      setState(() {
        _purchases.addAll(purchaseDetailsList);
        _listenToPurchaseUpdated(purchaseDetailsList);
      });
    }, onDone: () {
      _subscriptions.cancel();
    }, onError: (error) {
      //handle errors here
      _subscriptions.cancel();
    });

    /// initialization of the in app purchase
    _initialize();

    super.initState();
  }

  @override
  void dispose() {
    _subscriptions.cancel();
    super.dispose();
  }

  void _initialize() async {
    /// check if in app purchase is available or not
    _isAvailable = await _inAppPurchase.isAvailable();

    if (_isAvailable) {
      // get products from the play store
      List<ProductDetails> products = await _getProducts();
      _products = products;
      setState(() {});
    }


 void _listenToPurchaseUpdated(List<PurchaseDetails> purchaseDetailsList) {
    purchaseDetailsList.forEach((PurchaseDetails purchaseDetails) async {
      switch (purchaseDetails.status) {
        case PurchaseStatus.pending:
          // Purchase pending show UI
          break;

        case PurchaseStatus.canceled:
          // Purchase canceled show UI
          break;

        case PurchaseStatus.restored:
          // In this point need the purchase token
          // Purchase restored show UI
          break;

        case PurchaseStatus.error:
          // Purchase pending show UI
          break;

        case PurchaseStatus.purchased:
          print("PURCHASE COMPLETED");
          break;

        default:
          break;
      }

      if (purchaseDetails.pendingCompletePurchase) {
        await _inAppPurchase.completePurchase(purchaseDetails);
      }
    });
  }

  Future<List<ProductDetails>> _getProducts() async {
    Set<String> ids = Set.from(_productIds);
    ProductDetailsResponse response =
        await _inAppPurchase.queryProductDetails(ids);
    return response.productDetails;
  }


// This function is used for subscribe the topic
void _subscribe({@required ProductDetails product}) {
    final PurchaseParam purchaseParam = PurchaseParam(productDetails: product);
    _inAppPurchase.buyNonConsumable(
      purchaseParam: purchaseParam,
    );
  }
  }

**Comment by darshankawar** [comment]
@joy-eiosys 
Can you upgrade to latest plugin version which has lot of [changes](https://pub.dev/packages/in_app_purchase/changelog#200) / updates made and also the framework version to latest (stable 2.8.0) and try again to see if it helps ?

**Comment by darshankawar** [comment]
@joy-eiosys 
Could be the case that on Android there's a way to consume the token after purchase is completed, but may not be in flutter library. Are you saying that once a purchase is completed or restored, you don't get any token for it ?

**Comment by joy-eiosys** [comment]
Yes, I'm not getting any purchase tokens after completing the purchase. Only getting this JSON: 

{
"purchaseID":"Combination of string and int",
"productID":"1",
"transactionDate":"2021-12-15",
"verificationData":{
"localVerificationData":"String",
"serverVerificationData":"String",
"source":"String"
}
}

**Comment by darshankawar** [comment]
Keeping it open for further insights on this from team. I guess the provision isn't there yet in the library.

/cc @mvanbeusekom

**Comment by joy-eiosys** [comment]
Can "purchaseID" parameter be used for the token I mentioned earlier?

**Comment by JojoGogo9** [comment]
I'm seeing this also on the latest version. After a non consumable purchase is used, I don't receive anything from the purchase stream after the initial stream. Maybe this is expected, but it seems like it used to pass me all the past receipts. This is on Android, I haven't tested it on iOS yet.
